qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=cyclic-import
+
+"""
+=========
+Pulse IR
+=========
+
+"""
+
+from __future__ import annotations
+from typing import List, Callable
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from qiskit.pulse.exceptions import PulseError
+
+from qiskit.pulse.transforms import AlignmentKind, AlignSequential
+from qiskit.pulse.instructions import Instruction
+
+
+class IrElement(ABC):
+    """Base class for Pulse IR elements"""
+
+    @property
+    @abstractmethod
+    def initial_time(self) -> int | None:
+        """Return the initial time of the element"""
+        pass
+
+    @property
+    @abstractmethod
+    def duration(self) -> int | None:
+        """Return the duration of the element"""
+        pass
+
+    @abstractmethod
+    def shift_initial_time(self, value: int):
```

Comment:
I think this must be implemented by a pass.

Reply:
Let's leave it in at the moment, and remove it later if we decide it's better to have it as a pass. I think during scheduling this is something we might have to do often, so it might make sense to have it as a method.